File: python_backup/src/aigpt/relationship.py
"""Relationship tracking system with irreversible damage"""

import json
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, Optional
import logging

from .models import Relationship, RelationshipStatus


class RelationshipTracker:
    """Tracks and manages relationships with users"""
    
    def __init__(self, data_dir: Path):
        self.data_dir = data_dir
        self.relationships_file = data_dir / "relationships.json"
        self.relationships: Dict[str, Relationship] = {}
        self.logger = logging.getLogger(__name__)
        self._load_relationships()
    
    def _load_relationships(self):
        """Load relationships from persistent storage"""
        if self.relationships_file.exists():
            with open(self.relationships_file, 'r', encoding='utf-8') as f:
                data = json.load(f)
                for user_id, rel_data in data.items():
                    self.relationships[user_id] = Relationship(**rel_data)
    
    def _save_relationships(self):
        """Save relationships to persistent storage"""
        data = {
            user_id: rel.model_dump(mode='json')
            for user_id, rel in self.relationships.items()
        }
        with open(self.relationships_file, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, default=str)
    
    def get_or_create_relationship(self, user_id: str) -> Relationship:
        """Get existing relationship or create new one"""
        if user_id not in self.relationships:
            self.relationships[user_id] = Relationship(user_id=user_id)
            self._save_relationships()
        return self.relationships[user_id]
    
    def update_interaction(self, user_id: str, delta: float) -> Relationship:
        """Update relationship based on interaction"""
        rel = self.get_or_create_relationship(user_id)
        
        # Check if relationship is broken (irreversible)
        if rel.is_broken:
            self.logger.warning(f"Relationship with {user_id} is broken. No updates allowed.")
            return rel
        
        # Check daily limit
        if rel.last_interaction and rel.last_interaction.date() == datetime.now().date():
            if rel.daily_interactions >= rel.daily_limit:
                self.logger.info(f"Daily interaction limit reached for {user_id}")
                return rel
        else:
            rel.daily_interactions = 0
        
        # Update interaction counts
        rel.daily_interactions += 1
        rel.total_interactions += 1
        rel.last_interaction = datetime.now()
        
        # Update score with bounds
        old_score = rel.score
        rel.score += delta
        rel.score = max(0.0, min(200.0, rel.score))  # 0-200 range
        
        # Check for relationship damage
        if delta < -10.0:  # Significant negative interaction
            self.logger.warning(f"Major relationship damage with {user_id}: {delta}")
            if rel.score <= 0:
                rel.is_broken = True
                rel.status = RelationshipStatus.BROKEN
                rel.transmission_enabled = False
                self.logger.error(f"Relationship with {user_id} is now BROKEN (irreversible)")
        
        # Update relationship status based on score
        if not rel.is_broken:
            if rel.score >= 150:
                rel.status = RelationshipStatus.CLOSE_FRIEND
            elif rel.score >= 100:
                rel.status = RelationshipStatus.FRIEND
            elif rel.score >= 50:
                rel.status = RelationshipStatus.ACQUAINTANCE
            else:
                rel.status = RelationshipStatus.STRANGER
            
            # Check transmission threshold
            if rel.score >= rel.threshold and not rel.transmission_enabled:
                rel.transmission_enabled = True
                self.logger.info(f"Transmission enabled for {user_id}!")
        
        self._save_relationships()
        return rel
    
    def apply_time_decay(self):
        """Apply time-based decay to all relationships"""
        now = datetime.now()
        
        for user_id, rel in self.relationships.items():
            if rel.is_broken or not rel.last_interaction:
                continue
            
            # Calculate days since last interaction
            days_inactive = (now - rel.last_interaction).days
            
            if days_inactive > 0:
                # Apply decay
                decay_amount = rel.decay_rate * days_inactive
                old_score = rel.score
                rel.score = max(0.0, rel.score - decay_amount)
                
                # Update status if score dropped
                if rel.score < rel.threshold:
                    rel.transmission_enabled = False
                
                if decay_amount > 0:
                    self.logger.info(
                        f"Applied decay to {user_id}: {old_score:.2f} -> {rel.score:.2f}"
                    )
        
        self._save_relationships()
    
    def get_transmission_eligible(self) -> Dict[str, Relationship]:
        """Get all relationships eligible for transmission"""
        return {
            user_id: rel
            for user_id, rel in self.relationships.items()
            if rel.transmission_enabled and not rel.is_broken
        }

File: src/relationship.rs
use std::collections::HashMap;
use serde::{Deserialize, Serialize};
use anyhow::{Result, Context};
use chrono::{DateTime, Utc};

use crate::config::Config;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Relationship {
    pub user_id: String,
    pub score: f64,
    pub threshold: f64,
    pub status: RelationshipStatus,
    pub total_interactions: u32,
    pub positive_interactions: u32,
    pub negative_interactions: u32,
    pub transmission_enabled: bool,
    pub is_broken: bool,
    pub last_interaction: Option<DateTime<Utc>>,
    pub last_transmission: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub daily_interaction_count: u32,
    pub last_daily_reset: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum RelationshipStatus {
    New,
    Acquaintance,
    Friend,
    CloseFriend,
    Broken,
}

impl std::fmt::Display for RelationshipStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            RelationshipStatus::New => write!(f, "new"),
            RelationshipStatus::Acquaintance => write!(f, "acquaintance"),
            RelationshipStatus::Friend => write!(f, "friend"),
            RelationshipStatus::CloseFriend => write!(f, "close_friend"),
            RelationshipStatus::Broken => write!(f, "broken"),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RelationshipTracker {
    relationships: HashMap<String, Relationship>,
    config: Config,
}

impl RelationshipTracker {
    pub fn new(config: &Config) -> Result<Self> {
        let relationships = Self::load_relationships(config)?;
        
        Ok(RelationshipTracker {
            relationships,
            config: config.clone(),
        })
    }
    
    pub fn get_or_create_relationship(&mut self, user_id: &str) -> &mut Relationship {
        let now = Utc::now();
        
        self.relationships.entry(user_id.to_string()).or_insert_with(|| {
            Relationship {
                user_id: user_id.to_string(),
                score: 0.0,
                threshold: 10.0, // Default threshold for transmission
                status: RelationshipStatus::New,
                total_interactions: 0,
                positive_interactions: 0,
                negative_interactions: 0,
                transmission_enabled: false,
                is_broken: false,
                last_interaction: None,
                last_transmission: None,
                created_at: now,
                daily_interaction_count: 0,
                last_daily_reset: now,
            }
        })
    }
    
    pub fn process_interaction(&mut self, user_id: &str, sentiment: f64) -> Result<f64> {
        let now = Utc::now();
        let score_change;
        
        // Create relationship if it doesn't exist
        {
            let relationship = self.get_or_create_relationship(user_id);
            
            // Reset daily count if needed
            if (now - relationship.last_daily_reset).num_days() >= 1 {
                relationship.daily_interaction_count = 0;
                relationship.last_daily_reset = now;
            }
            
            // Apply daily interaction limit
            if relationship.daily_interaction_count >= 10 {
                return Ok(0.0); // No score change due to daily limit
            }
            
            // Store previous score for potential future logging
            
            // Calculate score change based on sentiment
            let mut base_score_change = sentiment * 0.5; // Base change
            
            // Apply diminishing returns for high interaction counts
            let interaction_factor = 1.0 / (1.0 + relationship.total_interactions as f64 * 0.01);
            base_score_change *= interaction_factor;
            score_change = base_score_change;
            
            // Update relationship data
            relationship.score += score_change;
            relationship.score = relationship.score.max(-50.0).min(100.0); // Clamp score
            relationship.total_interactions += 1;
            relationship.daily_interaction_count += 1;
            relationship.last_interaction = Some(now);
            
            if sentiment > 0.0 {
                relationship.positive_interactions += 1;
            } else if sentiment < 0.0 {
                relationship.negative_interactions += 1;
            }
            
            // Check for relationship breaking
            if relationship.score <= -20.0 && !relationship.is_broken {
                relationship.is_broken = true;
                relationship.transmission_enabled = false;
                relationship.status = RelationshipStatus::Broken;
            }
            
            // Enable transmission if threshold is reached
            if relationship.score >= relationship.threshold && !relationship.is_broken {
                relationship.transmission_enabled = true;
            }
        }
        
        // Update status based on score (separate borrow)
        self.update_relationship_status(user_id);
        
        self.save_relationships()?;
        
        Ok(score_change)
    }
    
    fn update_relationship_status(&mut self, user_id: &str) {
        if let Some(relationship) = self.relationships.get_mut(user_id) {
            if relationship.is_broken {
                return; // Broken relationships cannot change status
            }
            
            relationship.status = match relationship.score {
                score if score >= 50.0 => RelationshipStatus::CloseFriend,
                score if score >= 20.0 => RelationshipStatus::Friend,
                score if score >= 5.0 => RelationshipStatus::Acquaintance,
                _ => RelationshipStatus::New,
            };
        }
    }
    
    pub fn apply_time_decay(&mut self) -> Result<()> {
        let now = Utc::now();
        let decay_rate = 0.1; // 10% decay per day
        
        for relationship in self.relationships.values_mut() {
            if let Some(last_interaction) = relationship.last_interaction {
                let days_since_interaction = (now - last_interaction).num_days() as f64;
                
                if days_since_interaction > 0.0 {
                    let decay_factor = (1.0_f64 - decay_rate).powf(days_since_interaction);
                    relationship.score *= decay_factor;
                    
                    // Update status after decay
                    if relationship.score < relationship.threshold {
                        relationship.transmission_enabled = false;
                    }
                }
            }
        }
        
        // Update statuses for all relationships
        let user_ids: Vec<String> = self.relationships.keys().cloned().collect();
        for user_id in user_ids {
            self.update_relationship_status(&user_id);
        }
        
        self.save_relationships()?;
        Ok(())
    }
    
    pub fn get_relationship(&self, user_id: &str) -> Option<&Relationship> {
        self.relationships.get(user_id)
    }
    
    pub fn list_all_relationships(&self) -> &HashMap<String, Relationship> {
        &self.relationships
    }
    
    pub fn get_transmission_eligible(&self) -> HashMap<String, &Relationship> {
        self.relationships
            .iter()
            .filter(|(_, rel)| rel.transmission_enabled && !rel.is_broken)
            .map(|(id, rel)| (id.clone(), rel))
            .collect()
    }
    
    pub fn record_transmission(&mut self, user_id: &str) -> Result<()> {
        if let Some(relationship) = self.relationships.get_mut(user_id) {
            relationship.last_transmission = Some(Utc::now());
            self.save_relationships()?;
        }
        Ok(())
    }
    
    pub fn get_relationship_stats(&self) -> RelationshipStats {
        let total_relationships = self.relationships.len();
        let active_relationships = self.relationships
            .values()
            .filter(|r| r.total_interactions > 0)
            .count();
        let transmission_enabled = self.relationships
            .values()
            .filter(|r| r.transmission_enabled)
            .count();
        let broken_relationships = self.relationships
            .values()
            .filter(|r| r.is_broken)
            .count();
        
        let avg_score = if total_relationships > 0 {
            self.relationships.values().map(|r| r.score).sum::<f64>() / total_relationships as f64
        } else {
            0.0
        };
        
        RelationshipStats {
            total_relationships,
            active_relationships,
            transmission_enabled,
            broken_relationships,
            avg_score,
        }
    }
    
    fn load_relationships(config: &Config) -> Result<HashMap<String, Relationship>> {
        let file_path = config.relationships_file();
        if !file_path.exists() {
            return Ok(HashMap::new());
        }
        
        let content = std::fs::read_to_string(file_path)
            .context("Failed to read relationships file")?;
        
        let relationships: HashMap<String, Relationship> = serde_json::from_str(&content)
            .context("Failed to parse relationships file")?;
        
        Ok(relationships)
    }
    
    fn save_relationships(&self) -> Result<()> {
        let content = serde_json::to_string_pretty(&self.relationships)
            .context("Failed to serialize relationships")?;
        
        std::fs::write(&self.config.relationships_file(), content)
            .context("Failed to write relationships file")?;
        
        Ok(())
    }
    
    pub fn get_all_relationships(&self) -> Result<HashMap<String, RelationshipCompact>> {
        let mut result = HashMap::new();
        
        for (user_id, relationship) in &self.relationships {
            result.insert(user_id.clone(), RelationshipCompact {
                score: relationship.score,
                trust_level: relationship.score / 10.0, // Simplified trust calculation
                interaction_count: relationship.total_interactions,
                last_interaction: relationship.last_interaction.unwrap_or(relationship.created_at),
                status: relationship.status.clone(),
            });
        }
        
        Ok(result)
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct RelationshipStats {
    pub total_relationships: usize,
    pub active_relationships: usize,
    pub transmission_enabled: usize,
    pub broken_relationships: usize,
    pub avg_score: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct RelationshipCompact {
    pub score: f64,
    pub trust_level: f64,
    pub interaction_count: u32,
    pub last_interaction: DateTime<Utc>,
    pub status: RelationshipStatus,
}